flocker/control/_protocol.py

```diff
@@ -455,27 +458,52 @@ def _send_state_to_connections(self, connections):
         """
         configuration = self.configuration_service.get()
         state = self.cluster_state.as_deployment()
-        with LOG_SEND_CLUSTER_STATE(self.logger,
-                                    configuration=configuration,
+        with LOG_SEND_CLUSTER_STATE(configuration=configuration,
                                     state=state):
             with _caching_encoder.cache():
                 for connection in connections:
-                    action = LOG_SEND_TO_AGENT(self.logger, agent=connection)
-                    with action.context():
-                        # XXX If callRemote raises an exception, the loop won't
-                        # finish and the rest of the connections won't receive
-                        # the updated state.  Asynchronous exceptions aren't a
-                        # problem since they won't interrupt the loop (and they
-                        # shouldn't be allowed to).  No test coverage for
-                        # either of these cases.
-                        d = DeferredContext(connection.callRemote(
-                            ClusterStatusCommand,
-                            configuration=configuration,
-                            state=state,
-                            eliot_context=action
-                        ))
-                        d.addActionFinish()
-                        d.result.addErrback(lambda _: None)
+                    # XXX If callRemote raises an exception, the loop won't
+                    # finish and the rest of the connections won't receive
+                    # the updated state.  Asynchronous exceptions aren't a
+                    # problem since they won't interrupt the loop (and they
+                    # shouldn't be allowed to).  No test coverage for
+                    # either of these cases.
+                    try:
+                        (current_command, already_scheduled) = self._current_command_for_connection[
+                            connection
+                        ]
+                    except KeyError:
+                        current_command = self._update_connection(
+                            connection, configuration, state,
+                        )
+                        self._current_command_for_connection[connection] = (current_command, False)
+
+                        def finished_update(ignored, connection):
+                            del self._current_command_for_connection[connection]
+                        current_command.addCallback(finished_update, connection)
+                    else:
+                        if already_scheduled:
+                            AGENT_UPDATE_ELIDED(agent=connection).write()
+                        else:
```

Comment:
# Schedule another state change message.
# When the outstanding message is eventually acknowledged, the latest state and configuration will be immediately sent to this agent.